sn_logging/src/metrics.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2023 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use serde::Serialize;
+use std::time::Duration;
+use sysinfo::{self, CpuExt, NetworkExt, Pid, PidExt, ProcessExt, System, SystemExt};
+use tracing::trace;
+
+const UPDATE_INTERVAL: Duration = Duration::from_secs(5);
+const TO_MB: f32 = 1_000_000.0;
+
+// The following Metrics are collected and logged
+#[derive(Debug, Serialize)]
+#[allow(dead_code)]
+struct Metrics {
+    // Number of threads in the system
+    physical_cpu_threads: usize,
+    // Percentage of CPU used
+    cpu_usage_percent: f32,
+    // Total system memory in MBytes
+    total_memory_mb: f32,
+    // RAM used in MBytes
+    memory_used_mb: f32,
+    // Percentage of RAM used
+    memory_usage_percent: f32,
+    // Network metrics is None if the default network interface cannot be found
+    network: Option<NetworkMetrics>,
+    // Process metrics is None if the Pid for the process is invalid
+    process: Option<ProcessMetrics>,
+}
+
+#[derive(Debug, Serialize)]
+#[allow(dead_code)]
+struct NetworkMetrics {
+    // The default network interface
+    interface_name: String,
+    // Bytes received during UPDATE_INTERVAL
+    bytes_received: u64,
+    // Bytes transmitted during UPDATE_INTERVAL
+    bytes_transmitted: u64,
+    // The total MBytes received through the interface
+    total_mb_received: f32,
+    // The total MBytes transmitted through the interface
+    total_mb_transmitted: f32,
```

Comment:
number of messages received & transmitted as well?

Reply:
Do you mean the `safenode` related internal metrics? Not sure of a clean approach to do that while also being feature gated. :thinking: Passing a channel would be messy?

Reply:
Ah, yeah, I realized that's an `internal affair`, leave it then.
sry for the disturb